From ee3f3faa5f22fa01ed2fe899280cfdc0ae88702f Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Wed, 10 Aug 2022 13:43:41 -0700
Subject: [PATCH 05/12] ath12k: do not parse desctiptor that will not be
 handled

Incase of HAL_WBM_REL_HTT_TX_COMP_STATUS_MEC_NOTIFY, driver does
not process this anyway. Hence avoid full desctriptor parsing for
this case.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 13044e5b7156..9562b9a7c965 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -749,6 +749,22 @@ static inline void ath12k_dp_tx_status_parse(struct ath12k_base *ab,
 	ts->flags = FIELD_GET(HAL_WBM_RELEASE_TX_INFO2_FIRST_MSDU ,desc->info2) | FIELD_GET(HAL_WBM_RELEASE_TX_INFO2_LAST_MSDU ,desc->info2);
 }
 
+static inline bool ath12k_dp_tx_completion_valid(struct hal_wbm_release_ring *desc)
+{
+	struct htt_tx_wbm_completion *status_desc;
+
+	if (FIELD_GET(HAL_WBM_COMPL_TX_INFO0_REL_SRC_MODULE, desc->info0) ==
+	    HAL_WBM_REL_SRC_MODULE_FW) {
+		status_desc = ((u8 *)desc) + HTT_TX_WBM_COMP_STATUS_OFFSET;
+
+		/* Dont consider HTT_TX_COMP_STATUS_MEC_NOTIFY */
+		if (FIELD_GET(HTT_TX_WBM_COMP_INFO0_STATUS, status_desc->info0) ==
+		    HAL_WBM_REL_HTT_TX_COMP_STATUS_MEC_NOTIFY)
+			return false;
+	}
+	return true;
+}
+
 void ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id)
 {
 	struct ath12k *ar;
@@ -778,6 +794,9 @@ void ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id)
 	ath12k_hal_srng_dst_invalidate_entry(ab, status_ring, valid_entries);
 
 	while ((desc = ath12k_hal_srng_dst_get_next_cache_entry(ab, status_ring))) {
+		if (!ath12k_dp_tx_completion_valid(desc))
+			continue;
+
 		memcpy(&tx_ring->tx_status[count],
 		       desc, sizeof(struct hal_wbm_release_ring));
 		count++;
-- 
2.17.1

